functions/go/set-labels/transformer/labels.go

```diff
@@ -99,78 +75,365 @@ func (p *LabelTransformer) Transform(objects fn.KubeObjects) error {
+			for _, nextVecObj := range vecObj.GetSlice("to") {
+				err := updateLabels(nextVecObj, podSelector, p.NewLabels, false, p.ResultCount)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// hasSpecSelector check if the resource contains struct SpecSelector, kind Service and ReplicationController has it
+func hasSpecSelector(o *fn.KubeObject) bool {
+	if o.IsGVK("", "v1", "Service") || o.IsGVK("", "v1", "ReplicationController") {
+		return true
+	}
+	return false
+}
+
+// setSelector set labels for all selectors, including spec selector map, spec selector LabelSelector, LabelSelector in JobTemplate, and podSelector in NetworkPolicy, and
+func (p *LabelTransformer) setSelector(o *fn.KubeObject) error {
+	if hasSpecSelector(o) {
+		fieldPath := FieldPath{"spec", "selector"}
+		if err := updateLabels(&o.SubObject, fieldPath, p.NewLabels, true, p.ResultCount); err != nil {
+			return err
+		}
+	}
+	if found, create := hasLabelSelector(o); found {
+		fieldPath := FieldPath{"spec", "selector", "matchLabels"}
+		if err := updateLabels(&o.SubObject, fieldPath, p.NewLabels, create, p.ResultCount); err != nil {
 			return err
 		}
-		// handle special case with slice
-		if o.IsGVK("apps", "", "StatefulSet") {
-			if o.GetMap("spec") != nil {
-				for _, vctObj := range o.GetMap("spec").GetSlice("volumeClaimTemplates") {
-					updatedLabels, err = updateLabels(vctObj, metaLabelsPath, p.NewLabels, true)
+	}
+	if hasJobTemplateSpec(o) {
+		fieldPath := FieldPath{"spec", "jobTemplate", "spec", "selector", "matchLabels"}
+		if err := updateLabels(&o.SubObject, fieldPath, p.NewLabels, false, p.ResultCount); err != nil {
+			return err
+		}
+	}
+	if hasNetworkPolicySpec(o) {
+		fieldPath := FieldPath{"spec", "podSelector", "matchLabels"}
+		if err := updateLabels(&o.SubObject, fieldPath, p.NewLabels, false, p.ResultCount); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// hasLabelSelector check if the resource contains struct LabelSelector, return (if the resource has LabelSelector, if the LabelSelector need to be created if not exist)
+func hasLabelSelector(o *fn.KubeObject) (bool, bool) {
+	if o.IsGVK("", "", "Deployment") || o.IsGVK("", "", "ReplicaSet") || o.IsGVK("", "", "DaemonSet") || o.IsGVK("apps", "", "StatefulSet") {
+		return true, true
+	}
+	if o.IsGVK("batch", "", "Job") || o.IsGVK("policy", "", "PodDisruptionBudget") {
+		return true, false
+	}
+	return false, false
+}
+
+// hasPodTemplateSpec check if the resource contains struct PodTemplateSpec, ReplicationController, Deployment, ReplicaSet, DaemonSet, StatefulSet, Job kind has it
+func hasPodTemplateSpec(o *fn.KubeObject) bool {
+	switch {
+	case o.IsGVK("", "v1", "ReplicationController"):
+		return true
+	case o.IsGVK("", "", "Deployment"):
+		return true
+	case o.IsGVK("", "", "ReplicaSet"):
+		return true
+	case o.IsGVK("", "", "DaemonSet"):
+		return true
+	case o.IsGVK("apps", "", "StatefulSet"):
+		return true
+	case o.IsGVK("batch", "", "Job"):
```

Comment:
Discussion point for @yuwenma, I think this would be a good use case for getting this information from the openapi data instead of having to hardcode all of these things.